language/move-compiler/src/to_bytecode/optimize/mod.rs

```diff
@@ -0,0 +1,39 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::parser::ast::FunctionName;
+use move_ir_types::ast::{self as IR};
+use std::collections::{BTreeSet, HashMap};
+
+mod remove_fallthrough_jumps;
+mod remove_unnecessary_store;
+mod remove_unused_locals;
+
+pub(crate) fn code(
+    _f: &FunctionName,
+    loop_heads: &BTreeSet<IR::BlockLabel_>,
+    locals: &mut Vec<(IR::Var, IR::Type)>,
+    blocks: &mut IR::BytecodeBlocks,
+) {
+    let mut changed = true;
+    while changed {
+        let fallthrough_or_block_removed = remove_fallthrough_jumps::code(loop_heads, blocks);
+        let store_removed = remove_unnecessary_store::code(blocks);
+        let local_removed = remove_unused_locals::code(locals, blocks);
+        changed = fallthrough_or_block_removed || store_removed || local_removed;
+    }
+}
```

Comment:
This might be overkill, but could be generalised, to make it easier to add future optimisations (this also folds in the per-optimization fixpoint into the outer fixpoint.

```rust
trait LocalOptimization {
  fn optimize(
      &self, 
      loop_heads: &BtreeSet<IR::BlockLabel_>,
      locals: &mut Vec<(IR::Var, IR::Type)>,
      blocks: &mut IR:BytecodeBlocks,
  ) -> bool;
}

let optimizations: Vec<Box<dyn LocalOptimization>> = vec![
    Box::new(NopStore),
    Box::new(DeadStore),
    Box::new(UnusedLocals),
    Box::new(Fallthrough),
    Box::new(EmptyBlocks),
];

let mut last_changed = 0;
for optimization in optimizations.iter().cycle() {
    // Input is the same as the last time this optimization ran, fixpoint reached.
    if last_changed > optimizations.len() {
        break;
    }

    if optimization.optimize(loop_heads, locals, blocks) {
        last_changed = 0;
    } else {
        last_changed += 1;
    }
} 
```

Reply:
I like it, love the `cycle`. It is a nice trick

Reply:
I also cleaned up the cfgir's optimizations using this pattern, thanks again!